integration/tests/identity_provider/identity_provider_test.go

```diff
@@ -0,0 +1,379 @@
+
+		By("reading Kubernetes resources")
+		list, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(list.Items).To(HaveLen(1))
+
+		secrets, err := clientset.CoreV1().Secrets(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(secrets.Items).NotTo(BeEmpty())
+
+		By("ensuring the client does not have write access")
+		_, err = clientset.CoreV1().ConfigMaps(metav1.NamespaceDefault).Create(context.TODO(), &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "testdata",
+			},
+			Data: map[string]string{
+				"key": "value",
+			},
+		}, metav1.CreateOptions{})
+
+		Expect(err).To(HaveOccurred())
+
+		By("disassociating the identity provider")
+		cmd = params.EksctlCmd.
+			WithArgs(
+				"disassociate",
+				"identityprovider",
+				"--config-file", "-",
+				"--wait",
+			).
+			WithStdin(strings.NewReader(identityProviderClusterConfig)).
+			WithoutArg("--region", params.Region).
+			WithTimeout(45 * time.Minute)
+
+		Expect(cmd).To(RunSuccessfully())
+
+	})
+
+})
+
+var _ = AfterSuite(func() {
+	if !params.SkipCreate && !params.SkipDelete {
+		params.DeleteClusters()
+	}
+	if cleanupCognitoResources != nil {
+		Expect(cleanupCognitoResources()).To(Succeed())
+	}
+})
+
+func createOIDCClientset(eksAPI eksiface.EKSAPI, o *OIDCConfig, clusterName string) (kubernetes.Interface, error) {
+	contextName := fmt.Sprintf("%s@%s", "test", clusterName)
+
+	cluster, err := eksAPI.DescribeCluster(&eks.DescribeClusterInput{
+		Name: aws.String(clusterName),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "describing cluster")
+	}
+
+	certData, err := base64.StdEncoding.DecodeString(*cluster.Cluster.CertificateAuthority.Data)
+	if err != nil {
+		return nil, errors.Wrap(err, "unexpected error decoding certificate authority data")
+	}
+
+	config := clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			clusterName: {
+				Server:                   *cluster.Cluster.Endpoint,
+				CertificateAuthorityData: certData,
+			},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			contextName: {
+				Cluster:  clusterName,
+				AuthInfo: contextName,
+			},
+		},
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			contextName: {
+				AuthProvider: &clientcmdapi.AuthProviderConfig{
+					Name: "oidc",
+					Config: map[string]string{
+						"client-id":      o.clientID,
+						"id-token":       o.idToken,
+						"refresh-token":  o.refreshToken,
+						"idp-issuer-url": o.idpIssuerURL,
+					},
+				},
+			},
+		},
+		CurrentContext: contextName,
+	}
+
+	clientConfig, err := clientcmd.NewDefaultClientConfig(config, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating default client config")
+	}
+
+	return kubernetes.NewForConfig(clientConfig)
+}
+
+func setupCognitoProvider(clusterName, region string) (*OIDCConfig, error) {
```

Comment:
Can we break this function apart a bit? It's one giant thing that does things in an order which read difficult together. If we break it down a bit, it's more readable, like:

setupCognitoProvider...
initializePools
createAdminStuff
done

That would read a lot more fluidly.

Reply:
Yeah, agreed. It was on my list but I forgot to do it.